### PR TITLE
[SIS-080] Library Backend

### LINKED ISSUE #80: [SIS-080] Library backend
> As a user of the Sistema Lesson Planner, I want to be able to view my lesson plans in the Library, as well as favorite them.

## Technical Information

Endpoints to implement: 
`getAllLessonPlans`
`favouriteLessonPlan` (new API) (watch the "u" in favourite) (we are a commonwealth nation here smh)

You will find these endpoints inside `LessonPlanService.js` under `/services/`. You can implement them following the example from the `deleteLessonPlan` endpoint; specifically, make sure that you're *not* calling the RNFS library directly in the function, but rather the required function from `Local.js` that will make the call to the library for you. Use the path constant from `constants.js` if you need access to a directory path constant as defined in the [documentation](https://github.com/itinance/react-native-fs#constants)!

Since you'll be working on the same file, `LessonPlanService.js` in tandem with @Lemiex and @kurtislaw, make sure to discuss with them for this ticket! We suggest starting a DM group chat together as you're starting your ticket. _Specifically, please discuss with each other how to implement the LessonPlan object specification in `/models/` because you'll be using it a lot in the service file!_

Questions on implementation? Missing functions in Local.js for RNFS calls that you need? DM the PLs!

## Local Storage Structure

In RNFS.DocumentDirectoryPath, establish 2 folders: `Default` and `Favourited`. Lesson Plans when first created should all go into the Default folder, and each lesson plan should create its own directory. 
> **Example:** `Default/LESSON_PLAN_NAME/`

In each lesson plan's directory, you'll have the .json file that details its contents, as well as its corresponding activity cards that the user attached to it.

When a lesson plan is favourited, move its directory into the `Favourited` folder. `Favourited` lesson plans are essentially pinned lesson plans, and should always be _displayed first_ in the UI, unless the user is sorting their lesson plans by name, or date, or whatever.

Things to consider and discuss with your ticket group (aka @Lemiex and @kurtislaw):
- How should we initialize the two directory paths `Default` and `Favourited`? Does RNFS automatically do it when we write to it with a path? Try it!
- What's the best way to return errors? How do you write more informative error messages?
- Are all moveFile() and related operations done recursively (a.k.a do all the contents of the directory move too?)

### getAllLessonPlanNames
**Description:** 
Return an array of all the lesson plan names that are currently in storage. Make sure to look inside _both_ the `Default` and `Favourited` directories, and return all the lesson plan names in the `Favourited` directory first in the array! If no lesson plans exist, error is thrown.

**Discussion:** 
Check the [wiki](https://github.com/uoftblueprint/sistema/wiki/Lesson-Plan-API) for more info, as well as the [Endpoint Reference](https://docs.google.com/document/d/1JdqTkQoUPN7ov9fGmf_vruavwfDfUzUdNY5I9AccnKQ/edit)!

### favouriteLessonPlan
**Description:** 
Given a LessonPlan object, move it into the `Favourited` directory. Be sure to check whether or not it exists already in the Default directory, and throw an error if it doesn't!

**Discussion:** 
Check the [wiki](https://github.com/uoftblueprint/sistema/wiki/Lesson-Plan-API) for more info, as well as the [Endpoint Reference](https://docs.google.com/document/d/1JdqTkQoUPN7ov9fGmf_vruavwfDfUzUdNY5I9AccnKQ/edit)! 

### unFavouriteLessonPlan
**Description:** 
Given a LessonPlan object, move it into the `Default` directory. check whether or not it exists already in the Favourited directory, and throw an error if it doesn't!

**Discussion:** 
Check the [wiki](https://github.com/uoftblueprint/sistema/wiki/Lesson-Plan-API) for more info, as well as the [Endpoint Reference](https://docs.google.com/document/d/1JdqTkQoUPN7ov9fGmf_vruavwfDfUzUdNY5I9AccnKQ/edit)! 

## Resources

Definitely think out your strategy first before writing any code. This ticket is in general very collaborative. Discuss with the Lesson Plan group: @Lemiex & @kurtislaw & @helenaglow 
[Official RNFS Documentation](https://github.com/itinance/react-native-fs)
[Testing RNFS](https://github.com/uoftblueprint/sistema/wiki/Testing-Local-Storage-Calls)

### PR DESCRIPTION
**Pull request by helenaglow**
# Description
Implementing unfavourite(), favourtie(), and getAllLessonPlanNames() endpoints. I have also added the mkdir method and more descriptive error messages in Local. I did check if the moved lesson plan's old path was deleted (it was), I just forgot to include it here.

Ticket #080
***Testing Instructions***

-> When you test, make sure you have wiped data on your emulator before each run, otherwise you are going to get some errors if you already made all the files.

**Add these functions and imports to App.js :**
First call:

      import { MAINDIRECTORY } from './src/services/constants';
      import {writeFile, makeDirectory, checkFileExists, readDirectory} from './src/services/routes/Local.js';
      import LessonPlanService from './src/services/LessonPlanService.js';
      
      async function mkDir(){
        await makeDirectory(MAINDIRECTORY + '/Default/');
        await makeDirectory(MAINDIRECTORY + '/Favourited/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson1/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson2/');
        await makeDirectory(MAINDIRECTORY + '/Favourited/lesson3/');
        await makeDirectory(MAINDIRECTORY + '/Favourited/lesson4/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson5/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson6/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson7/');
      }
      async function writeF(){
        await writeFile(MAINDIRECTORY + '/Default/lesson1/lessonplan1.txt', "hello");
        await writeFile(MAINDIRECTORY + '/Default/lesson2/lessonplan2.txt', "whatup");
        await writeFile(MAINDIRECTORY + '/Favourited/lesson3/lessonplan3.txt', "hey");
        await writeFile(MAINDIRECTORY + '/Favourited/lesson4/lessonplan4.txt', "gahhh");
        // await writeFile(MAINDIRECTORY + '/Default/lesson5/lessonplan5.txt', "gggg");
        // await writeFile(MAINDIRECTORY + '/Default/lesson6/lessonplan6.txt', "ffff");
        // await writeFile(MAINDIRECTORY + '/Default/lesson7/lessonplan7.txt', "hhhh");
      }
      
      async function checkExists(){
        await checkFileExists(MAINDIRECTORY + '/Default/');
        await checkFileExists(MAINDIRECTORY + '/Favourtied/');
        // await checkFileExists(MAINDIRECTORY + '/Default/lesson7/');
      }
     
      async function favourite(){
        await LessonPlanService.favouriteLessonPlan('lesson1');
        await LessonPlanService.unfavouriteLessonPlan('lesson4');
        // await LessonPlanService.favouriteLessonPlan('lesson6');
        // await checkFileExists(MAINDIRECTORY + '/Favourited/lesson6/');
        // await readDirectory(MAINDIRECTORY + '/Favourited/lesson6/')
      }
      
      async function listAllPlans(){
        let allplans = await LessonPlanService.getAllLessonPlanNames();
        console.log(`ALL PLANS: ${allplans}`);
      }
      
      async function calls(){
        mkDir()
        .then(() =>{
          checkExists()
          .then(() =>{
            writeF()
            .then(() => {
              favourite()
              .then(() =>
              listAllPlans())
            })
          })
        })
        .catch(err => {
          console.error(`errpr: ${err.message}`);
        });
      }

Second call (change code and apply changes to emulator):

      import { MAINDIRECTORY } from './src/services/constants';
      import {writeFile, makeDirectory, checkFileExists, readDirectory} from './src/services/routes/Local.js';
      import LessonPlanService from './src/services/LessonPlanService.js';
      
      async function mkDir(){
        // await makeDirectory(MAINDIRECTORY + '/Default/');
        // await makeDirectory(MAINDIRECTORY + '/Favourited/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson1/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson2/');
        // await makeDirectory(MAINDIRECTORY + '/Favourited/lesson3/');
        // await makeDirectory(MAINDIRECTORY + '/Favourited/lesson4/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson5/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson6/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson7/');
      }
      async function writeF(){
        // await writeFile(MAINDIRECTORY + '/Default/lesson1/lessonplan1.txt', "hello");
        // await writeFile(MAINDIRECTORY + '/Default/lesson2/lessonplan2.txt', "whatup");
        // await writeFile(MAINDIRECTORY + '/Favourited/lesson3/lessonplan3.txt', "hey");
        // await writeFile(MAINDIRECTORY + '/Favourited/lesson4/lessonplan4.txt', "gahhh");
        await writeFile(MAINDIRECTORY + '/Default/lesson5/lessonplan5.txt', "gggg");
        await writeFile(MAINDIRECTORY + '/Default/lesson6/lessonplan6.txt', "ffff");
        await writeFile(MAINDIRECTORY + '/Default/lesson7/lessonplan7.txt', "hhhh");
      }
      
      async function checkExists(){
        //await checkFileExists(MAINDIRECTORY + '/Default/');
        //await checkFileExists(MAINDIRECTORY + '/Favourtied/');
        await checkFileExists(MAINDIRECTORY + '/Default/lesson7/');
      }
      
      async function favourite(){
        // await LessonPlanService.favouriteLessonPlan('lesson1');
        // await LessonPlanService.unfavouriteLessonPlan('lesson4');
        await LessonPlanService.favouriteLessonPlan('lesson6');
        await checkFileExists(MAINDIRECTORY + '/Favourited/lesson6/');
        await readDirectory(MAINDIRECTORY + '/Favourited/lesson6/')
      }
      
      async function listAllPlans(){
        let allplans = await LessonPlanService.getAllLessonPlanNames();
        console.log(`ALL PLANS: ${allplans}`);
      }
      
      async function calls(){
        mkDir()
        .then(() =>{
          checkExists()
          .then(() =>{
            writeF()
            .then(() => {
              favourite()
              .then(() =>
              listAllPlans())
            })
          })
        })
        .catch(err => {
          console.error(`error: ${err.message}`);
        });
      }

**Then in MainNavigator add:**

    calls();

**Here are the test results, I got back:**
- First call: 

image

- Second call:

image

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# Checklist:

- [X] My PR title is formatted as [SIS-\<issue number\>] \<issue name\> (eg. [SIS-010] Create UI button)
- [X] I have linked my issue/ticket on the project board to this PR
- [X] I have sent my PR to #team-sistema-prs on Slack and requested two people to review
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (if required)
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by yuemily24** [comment]
image

File: src/services/routes/Local.js
Comment (yuemily24) [inline src/services/routes/Local.js:9]: Hi Helena, I think we'll need to rehaul the RNFS call here, sorry it will likely change the rest of your code but should be an easy fix. Essentially, we need to grab the last modified date of the file in order to sort the lesson plans in the Library by last edited (also display it in the [options menu)](https://www.figma.com/file/kSZduD0A5hk8sHmvqiC7DE/Sistema-Hi-Fidelity-Designs-%2B-Design-System?node-id=639%3A18072&t=x27PBltneYiqUkgE-0). This means, we need to change this RNFS call to [RNFS.readDir](https://github.com/itinance/react-native-fs#readdirdirpath-string-promisereaddiritem), which will return `Promise<ReadDirItem[]>`. 

Here's what you can follow along:
1. In `Local.js`, call `RNFS.readDir` and return `ReadDirItem[]` without any modifications.
2. Inside `LessonPlanService.js`, map the result and extract the info. For instance, 
```javascript
// where result is ReadDirItem[]
const lpInfo = result.map((dirItem: ReadDirItem) => {
    // do whatever manipulation, maybe like turn the date into DateTime object
    return { mtime: date,     // The last modified date of the file
                 name: string }     // The name of the item
})
```
You might even want to return the array type of `lpInfo` into a model in`model.js`, so that it's clear on the frontend what parameters are in the object you're getting. 
Also, don't worry about sorting the array right away, we'll do that on the front-end.

Comment (helenaglow) [reply src/services/routes/Local.js:9]: I fixed this, instead of changing readDirectory (since i use it in a different funciton), I added a new method in local for readDir called readDDirectory.

**Comment by helenaglow** [comment]
Hi hi! I updated my code! The front end and every thing works, but please let me know if you catch any issues. I will work those asap. (There is one little thing that I'm gonna test but it doesn't change much)

File: src/library/Library.js
Comment (ramyzhang) [inline src/library/Library.js]: is it possible for you to remove these here and just do it inside handleFavouriteChange via the following Promise syntax instead?
```
LessonPlanService.favouriteLessonPlan(name)
    .then(() => {
        // do something here 
    })
    .catch((err) => {
        console.error(blablablabla)
    }))
```

**Review by ramyzhang — CHANGES_REQUESTED** [review]
great job!!! just a small edit as mentioned in my comment, and it'll be ready to merge after you resolve the merge conflicts :~) thanks for your hard work implementing this, this is awesome!!!

File: src/library/Library.js
Comment (helenaglow) [reply src/library/Library.js]: Hi it should be fixed now!

Comment (ramyzhang) [reply src/library/Library.js]: perfect!!!

**Comment by helenaglow** [comment]
IT WORKS

**Pull request merged**
